src/modules/musig/keyagg_impl.h

```diff
@@ -0,0 +1,271 @@
+        secp256k1_scalar_set_int(r, 1);
+    } else {
+        secp256k1_musig_keyaggcoef_sha256(&sha);
+        secp256k1_sha256_write(&sha, pk_hash, 32);
+        secp256k1_fe_get_b32(buf, x);
+        secp256k1_sha256_write(&sha, buf, 32);
+        secp256k1_sha256_finalize(&sha, buf);
+        secp256k1_scalar_set_b32(r, buf, NULL);
+    }
+
+}
+
+/* Assumes both field elements x and second_pk_x are normalized. */
+static void secp256k1_musig_keyaggcoef(secp256k1_scalar *r, const secp256k1_keyagg_cache_internal *cache_i, secp256k1_fe *x) {
+    secp256k1_musig_keyaggcoef_internal(r, cache_i->pk_hash, x, &cache_i->second_pk_x);
+}
+
+typedef struct {
+    const secp256k1_context *ctx;
+    /* pk_hash is the hash of the public keys */
+    unsigned char pk_hash[32];
+    const secp256k1_xonly_pubkey * const* pks;
+    secp256k1_fe second_pk_x;
+} secp256k1_musig_pubkey_agg_ecmult_data;
+
+/* Callback for batch EC multiplication to compute pk_hash_0*P0 + pk_hash_1*P1 + ...  */
+static int secp256k1_musig_pubkey_agg_callback(secp256k1_scalar *sc, secp256k1_ge *pt, size_t idx, void *data) {
+    secp256k1_musig_pubkey_agg_ecmult_data *ctx = (secp256k1_musig_pubkey_agg_ecmult_data *) data;
+    int ret;
+    ret = secp256k1_xonly_pubkey_load(ctx->ctx, pt, ctx->pks[idx]);
+    /* pubkey_load can't fail because the same pks have already been loaded (and
+     * we test this) */
+    VERIFY_CHECK(ret);
+    secp256k1_musig_keyaggcoef_internal(sc, ctx->pk_hash, &pt->x, &ctx->second_pk_x);
+    return 1;
+}
+
+int secp256k1_musig_pubkey_agg(const secp256k1_context* ctx, secp256k1_scratch_space *scratch, secp256k1_xonly_pubkey *agg_pk, secp256k1_musig_keyagg_cache *keyagg_cache, const secp256k1_xonly_pubkey * const* pubkeys, size_t n_pubkeys) {
+    secp256k1_musig_pubkey_agg_ecmult_data ecmult_data;
+    secp256k1_gej pkj;
+    secp256k1_ge pkp;
+    size_t i;
+
+    VERIFY_CHECK(ctx != NULL);
+    if (agg_pk != NULL) {
+        memset(agg_pk, 0, sizeof(*agg_pk));
+    }
+    ARG_CHECK(pubkeys != NULL);
+    ARG_CHECK(n_pubkeys > 0);
+
+    ecmult_data.ctx = ctx;
+    ecmult_data.pks = pubkeys;
+    /* No point on the curve has an X coordinate equal to 0 */
+    secp256k1_fe_set_int(&ecmult_data.second_pk_x, 0);
+    for (i = 1; i < n_pubkeys; i++) {
+        secp256k1_ge pt;
+        if (!secp256k1_xonly_pubkey_load(ctx, &pt, pubkeys[i])) {
+            return 0;
+        }
+        if (secp256k1_memcmp_var(pubkeys[0], pubkeys[i], sizeof(*pubkeys[0])) != 0) {
+            ecmult_data.second_pk_x = pt.x;
+            break;
+        }
+    }
+
+    if (!secp256k1_musig_compute_pk_hash(ctx, ecmult_data.pk_hash, pubkeys, n_pubkeys)) {
+        return 0;
+    }
+    if (!secp256k1_ecmult_multi_var(&ctx->error_callback, scratch, &pkj, NULL, secp256k1_musig_pubkey_agg_callback, (void *) &ecmult_data, n_pubkeys)) {
+        /* The current implementation of ecmult_multi_var makes this code unreachable with tests. */
```

Comment:
Maybe this is overkill but if you really want to get test coverage here, something like this work?

- Rename `musig_pubkey_agg` to `musig_pubkey_agg_internal` and pass in the callback as a parameter
- Have `musig_pubkey_agg` call `musig_pubkey_agg_internal` with the proper callback
- To reach this line, call `musig_pubkey_agg_internal` with a test-only callback that returns 0

Reply:
True, not sure why I phrased it that way. Seems a bit overkill though since this can't be reached outside of tests at all - unless ecmult_multi's behavior changes. I was mainly noting this because the goal is to get full branch coverage and this branch would stick out in gcov.

Reply:
I tried to rephrase the comment to remove "impossibe [...] in tests"